src/bootstrap/format.rs

```diff
@@ -1,7 +1,196 @@
 //! Runs rustfmt on the repository.
 
-use crate::Build;
+use {
+    crate::{util, Build, Crate},
+    glob::glob,
+    serde::Deserialize,
+    std::{
+        collections::BTreeSet,
+        fs,
+        path::{Path, PathBuf},
+        process::Command,
+    },
+};
 
 pub fn format(build: &Build, check: bool) {
-    unimplemented!("do the format command... {:?} {:?}", build.src, check);
+    let target = std::env::var("BUILD").expect("BUILD var is set by bootstrap.py");
+    let rustfmt_path = build.initial_rustc.with_file_name(util::exe("rustfmt", &target));
+    assert!(rustfmt_path.is_file(), "{} not a file", rustfmt_path.display());
+
+    let ignores = RustfmtIgnore::read(build);
+    let mut cmd = Command::new(rustfmt_path);
+    cmd.current_dir(&build.src);
+    ignores.add_args(&mut cmd);
+
+    if check {
+        cmd.arg("--check");
+    }
+
+    // the rest of the arguments are just the file entrypoints into the tree
```

Comment:
We'll want to pass `--edition 2018` I think as well, right?

Reply:
Apparently cargo-fmt applies that for us when the root crate is `edition = 2018` in Cargo.toml.